File: w10-mean/d2-drills-bfs/js/tree.js
// This object type represents a tree
// where each node has an array of children.

var Tree = function (key, children){
	this.key = key; 	
	this.children = children || [];	// an array, holds the Trees that we can go to directly from here
};

// Searches through all nodes of the tree, spreading 
// outward from the root. Looks for any node with key equal
// to the targetKey param. Returns null if no such node is found.
Tree.prototype.breadthFirstSearch = function(targetKey){
	return null;
};


// Returns an array containing all nodes for which 
// the selectionFunction returns true.
Tree.prototype.breadthFirstSearchAll = function(selectionFunction){
	return [];
};



var myTree = new TreeNode('A'),
	bNode = new TreeNode('B'),
	cNode = new TreeNode('C'),
	dNode = new TreeNode('D'),
	eNode = new TreeNode('E'),
	fNode = new TreeNode('F'),
	gNode = new TreeNode('G');

myTree.children.push(bNode, cNode);
bNode.children.push(dNode, eNode, fNode);
eNode.children.push(gNode);

//          D 
//        /
//     B  -- E -- G
//   /	  \
// A        F
//   \
//     C 

console.log("-- expect node E --");
console.log(myTree.breadthFirstSearch('E'));  
// queue:  [ 'A' ]
// queue:  [ 'B', 'C' ]
// queue:  [ 'C', 'D', 'E', 'F' ]
// queue:  [ 'D', 'E', 'F' ]
// queue:  [ 'E', 'F' ]
// { key: 'E', children: [ { key: 'G', children: [] } ] }

console.log("-- expect node G --");
console.log(myTree.breadthFirstSearch('G'));  
// queue:  [ 'A' ]
// queue:  [ 'B', 'C' ]
// queue:  [ 'C', 'D', 'E', 'F' ]
// queue:  [ 'D', 'E', 'F' ]
// queue:  [ 'E', 'F' ]
// queue:  [ 'F', 'G' ]
// queue:  [ 'G' ]
// { key: 'G', children: [] }


console.log("-- expect null --");
console.log(myTree.breadthFirstSearch('H')); 
// queue:  [ 'A' ]
// queue:  [ 'B', 'C' ]
// queue:  [ 'C', 'D', 'E', 'F' ]
// queue:  [ 'D', 'E', 'F' ]
// queue:  [ 'E', 'F' ]
// queue:  [ 'F', 'G' ]
// queue:  [ 'G' ]
// null


console.log("-- expect array with nodes E and F --");
console.log(myTree.breadthFirstSearchAll(function(node){
	return node.key > 'D';
}));
// queue:  [ 'A' ]
// queue:  [ 'B', 'C' ]
// queue:  [ 'C', 'D', 'E', 'F' ]
// queue:  [ 'D', 'E', 'F' ]
// queue:  [ 'E', 'F' ]
// queue:  [ 'F' ]
// [ { key: 'E', children: [ [Object] ] },
//   { key: 'F', children: [] } ]


console.log("-- expect empty array --");
console.log(myTree.breadthFirstSearchAll(function(node){
	return node.username == "Bob";
}));
// queue:  [ 'A' ]
// queue:  [ 'B', 'C' ]
// queue:  [ 'C', 'D', 'E', 'F' ]
// queue:  [ 'D', 'E', 'F' ]
// queue:  [ 'E', 'F' ]
// queue:  [ 'F', 'G' ]
// queue:  [ 'G' ]
// []

File: w10-mean/d2-drills-bfs/rb/tree.rb
##
# This class represents a tree
# where each node has an array of children.

class Tree

  ##
  # reads the key of the tree's root
  attr_reader :key

  ## 
  # sets the key of the tree's root node
  # unless intended new key is nil
  # (to remove a node, set the entire node to nil)
  def key=(new_key)
    if !new_key.nil?
      @key = new_key
    end
    @key
  end

  ##
  # sets or reads the array of children for this instance
  attr_accessor :children


  ##
  # Creates a new tree with the root node key specified by 
  # the +key+ param as the @key for the instance
  def initialize(key)
    @key = key
    @children = []
  end

  ##
  # Searches through all nodes of the tree, spreading 
  # outward from the root. Looks for any node with key equal
  # to the +target_key+ param. Returns nil if no such node is found.
  def breadth_first_search(target_key)
    nil
  end

end

my_tree = Tree.new('A')
b_node = Tree.new('B')
c_node = Tree.new('C')
d_node = Tree.new('D')
e_node = Tree.new('E')
f_node = Tree.new('F')
g_node = Tree.new('G')

my_tree.children << b_node
my_tree.children << c_node
b_node.children << d_node
b_node.children << e_node
b_node.children << f_node
e_node.children << g_node

#          D 
#        /
#     B  -- E -- G
#   /   \
# A        F
#   \
#     C 



puts "\n-- expect node E --"
e_result = my_tree.breadth_first_search('E')
p e_result
# ["A"]
# ["B", "C"]
# ["C", "D", "E", "F"]
# ["D", "E", "F"]
# ["E", "F"]
# #<Tree:0x007fee51891dd0 @key="E", @children=[#<Tree:0x007fee51891ce0 @key="G", @children=[]>]>

puts "\n-- expect node G --"
g_result = my_tree.breadth_first_search('G')
p g_result
# ["A"]
# ["B", "C"]
# ["C", "D", "E", "F"]
# ["D", "E", "F"]
# ["E", "F"]
# ["F", "G"]
# ["G"]
# #<Tree:0x007fe77c29dae0 @key="G", @children=[]>


puts "\n-- expect nil -- "
h_result = my_tree.breadth_first_search('H')
p h_result
# ["A"]
# ["B", "C"]
# ["C", "D", "E", "F"]
# ["D", "E", "F"]
# ["E", "F"]
# ["F", "G"]
# ["G"]
# nil

